sherpa/astro/utils/tests/test_astro_utils_unit.py

```diff
@@ -220,3 +224,269 @@ def test_do_group_single_group(func, expected):
 
     ans = do_group([6, 4, 2, 1, -1, 3], [1, -1, -1, -1, -1, -1], func)
     assert ans == pytest.approx([expected])
+
+
+def make_data(data_class):
+    """Create a test data object of the given class.
+
+    Using a string means it is easier to support the various
+    PHA "types" - eg basic, grouping, grouping+quality
+    """
+
+    x0 = np.asarray([1, 3, 7, 12])
+    y = np.asarray([2, 3, 4, 5])
+    if data_class == "1d":
+        return Data1D('x1', x0, y)
+
+    if data_class == "1dint":
+        return Data1DInt('xint1', x0, np.asarray([3, 5, 8, 15]), y)
+
+    chans = np.arange(1, 5)
+    if data_class == "pha":
+        return DataPHA('pha', chans, y)
+
+    # These are a special case to check out grpuping and quality handling.
+    grp = np.asarray([1, -1, 1, 1])
+    qual = np.asarray([0, 0, 2, 0])
+    pha = DataPHA('pha', chans, y, grouping=grp, quality=qual)
+    if data_class == "grp":
+        return pha
+
+    if data_class == "qual":
+        pha.ignore_bad()
+        return pha
+
+    x0 = np.asarray([1, 2, 3] * 2)
+    x1 = np.asarray([1, 1, 1, 2, 2, 2])
+    y = np.asarray([2, 3, 4, 5, 6, 7])
+    if data_class == "2d":
+        return Data2D('x2', x0, x1, y, shape=(2, 3))
+
+    if data_class ==  "2dint":
+        return Data2DInt('xint2', x0, x1, x0 + 1, x1 + 1, y, shape=(2, 3))
+
+    if data_class == "img":
+        return DataIMG('img', x0, x1, y, shape=(2, 3))
+
+    if data_class == "imgint":
+        return DataIMGInt('imgi', x0, x1, x0 + 1, x1 + 1, y, shape=(2, 3))
+
+    assert False
+
+
+@pytest.mark.parametrize("data_class", ["1d", "1dint", "pha", "grp", "qual"])
+def test_calc_data_sum_invalid_range(data_class):
+    """lo > hi"""
+
+    data = make_data(data_class)
+    with pytest.raises(IOErr) as err:
+        utils.calc_data_sum(data, 10, 2)
+
+    # TODO: This error message is not great.
+    #
+    assert str(err.value) == "the energy range is not consistent, 10 !< 2"
+
+
+@pytest.mark.parametrize("data_class", ["1d", "1dint", "pha", "grp", "qual"])
+def test_calc_data_sum_no_range(data_class):
+    """Call calc_data_sum(data)"""
+
+    data = make_data(data_class)
+    assert utils.calc_data_sum(data) == 14
```

Comment:
Mmm. I would have expected the answer to be different if I `ignore_bad` first, that is in the "qual" case. Do we really want to sum up known bad data? In this example, it's a reasonable number, but I practice, I would expect that bad data just as well could be inf. If that's current behavior and this is a just a regression test, I suggest adding at least a note or maybe split the qual in to a separate test with ` assert utils.calc_data_sum(data) == 10` and xfail?

Or maybe I'm just not understanding what `calc_data_sum` is supposed to be doig.

Reply:
So, this is the second point of my response above at https://github.com/sherpa/sherpa/pull/1440#discussion_r806262758 

This is not intended to be a PR that fixes all the things. The tests are written to test the current behavior. I really want to see them fail when we get around to fixing the quality handling, but do not want that to be "a subtle" change of x to X in the test output, but a "nope, this test has really failed because the code is now doing something different", as I am getting less hopeful that I'll ever get to fix the quality code.